lib/shared/require-or-import.js

```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+var importESM;
+try {
+  importESM = new Function('id', 'return import(id);');
+} catch (e) {
+  importESM = null;
+}
+
+function requireOrImport(id, callback) {
+  var err = null;
+  var cjs = null;
+  try {
+    cjs = require(id);
+  } catch (e) {
+    if (importESM && e.code === 'ERR_REQUIRE_ESM') {
+      importESM(id).then(function(esm) { callback(null, esm); }, callback);
+      return;
+    }
+    err = e;
+  }
+  process.nextTick(callback.bind(null, err, cjs));
```

Comment:
This removes the observable difference in `this`, `arguments.length` and `arguments[1]` when `import(…)` is used instead of `require(…)` and an error occurs:

```suggestion
  var cjs;
  try {
    cjs = require(id);
  } catch (err) {
    if (importESM && err.code === 'ERR_REQUIRE_ESM') {
      importESM(id).then(function(esm) { callback(null, esm); }, callback);
      return;
    }
    process.nextTick(function() {
      callback(err);
    });
    return;
  }
  process.nextTick(function () {
    callback(null, cjs);
  });
```

---

### Current:

| callback parameter | `require(…)`      | `import(…)`
| ------------------ | ------------      | -----------
| `this`             | `null`            | `undefined`
| `error`            | `Error \| null`   | `Error \| null`
| `imports`          | `exports \| null` | `exports \| undefined`
| `arguments.length` | `2`               | `1 \| 2`

### With this:

| callback parameter | `require(…)`           | `import(…)`
| ------------------ | ------------           | -----------
| `this`             | `undefined`            | `undefined`
| `error`            | `Error \| null`        | `Error \| null`
| `imports`          | `exports \| undefined` | `exports \| undefined`
| `arguments.length` | `1 \| 2`               | `1 \| 2`

Reply:
Thanks for pointing this out, but I wonder if it's worth the added complexity and duplication required to address this, in particular since this is an internal API. (FWIW, if it wouldn't be for Node.js 0.10 support, I would just create and return a new `Promise` here.) @phated, what do you think?

Reply:
I think the duplication is probably alright for this release. We are definitely switching everything to promises for the next major, but I really like that we can "backport" this to support our current release stream. That's why I'm so excited you put this together!

Reply:
Sure, I'm not suggesting to drop Node.js 0.10 support for this release. But I'm wondering whether you agree with @Exe-Boss or if you think my current implementation is fine? I mean it works, and there is no way a user can even tell the difference since it is an internal API. So I'm kinda reluctant to add redundant logic just for some hypothetically case. But your call. :)

Reply:
Sorry, I haven't fully reviewed this yet, so I was just responding to the questions.

I'll keep this in mind when I review.

Reply:
I partially addressed this now, except for the `arguments.length` case as this would require additional boilerplate (and I rather not bloat the code for a purely theoretical case).